controller/src/controller/storage/kubernetes/fsadmin.py

```diff
@@ -0,0 +1,193 @@
+
+from ...config import FSAdminConfig
+from ...models.domain.fsadmin import FSAdminObjects
+from ...models.domain.kubernetes import PodPhase
+from ...timeout import Timeout
+from ..metadata import MetadataStorage
+from .deleter import PersistentVolumeClaimStorage
+from .pod import PodStorage
+
+__all__ = ["FSAdminStorage"]
+
+
+class FSAdminStorage:
+    """Kubernetes storage layer for a filesystem admin pod.
+
+    Parameters
+    ----------
+    config
+        Configuration for fsadmin environment.
+    metadata_storage
+        Holds namespace information.
+    api_client
+        Kubernetes API client.
+    logger
+        Logger to use.
+
+    Notes
+    -----
+    This class isn't strictly necessary; instead, the fsadmin service
+    could call the storage layers for individual Kubernetes objects
+    directly. Even though there are not many objects, having a wrapper layer
+    might be easier to follow.
+    """
+
+    def __init__(
+        self,
+        *,
+        config: FSAdminConfig,
+        metadata_storage: MetadataStorage,
+        api_client: ApiClient,
+        logger: BoundLogger,
+    ) -> None:
+        self._config = config
+        self._logger = logger
+        self._metadata = metadata_storage
+        self._pod = PodStorage(api_client, logger)
+        self._pvc = PersistentVolumeClaimStorage(api_client, logger)
+        self._start_time: datetime.datetime | None = None
+
+    async def create(
+        self, objects: FSAdminObjects, timeout: Timeout
+    ) -> datetime.datetime:
+        """Create all of the Kubernetes objects for an fsadmin instance.
+
+        Create the objects in Kubernetes and then wait for the fsadmin pod
+        to start.
+
+        Returns the time at which the pod went into Running phase.
+
+        Parameters
+        ----------
+        objects
+            Kubernetes objects making up the fsadmin environment.
+        timeout
+            How long to wait for the fsadmin pod to start.
+
+        Returns
+        -------
+        datetime.datetime
+
+        Raises
+        ------
+        KubernetesError
+            Raised if there is some failure in a Kubernetes API call.
+        TimeoutError
+            Raised if fsadmin is not ready before the provided timeout expires.
+
+        Notes
+        -----
+        This is conceptually similar to the fileserver, but since it's a
+        cluster-wide singleton, if the objects all exist, that's fine.  We
+        don't try to delete and recreate, but use the existing fsadmin
+        instance.
+        """
+        ns = self._metadata.namespace
+        for pvc in objects.pvcs:
+            await self._pvc.create(ns, pvc, timeout)
+        await self._pod.create(ns, objects.pod, timeout)
+
+        # Wait for the pod to start.
+        await self._pod.wait_for_phase(
+            objects.pod.metadata.name,
+            ns,
+            until_not={PodPhase.UNKNOWN, PodPhase.PENDING},
+            timeout=timeout,
+        )
```

Comment:
It would be useful if `wait_for_phase` returned the `V1PodStatus` because then you'd have everything you needed already. Ah well, it's a trade-off for how it's used elsewhere. Since you have the status, though, you may as well put that into the `FSAdminStatus` as well.